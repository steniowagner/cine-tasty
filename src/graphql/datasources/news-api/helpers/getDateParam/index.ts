const getDateParam = (): string => {
  const today = new Date();

  const month = (today.getMonth() + 1).toString().padStart(2, '0');
  const year = today.getFullYear();
  const day = today
    .getDate()
    .toString()
    .padStart(2, '0');

  return `${year}-${month}-${day}`;
};

export default getDateParam;
