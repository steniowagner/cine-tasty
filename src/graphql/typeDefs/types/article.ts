import { gql } from 'apollo-server';

export default gql`
  enum ArticleLanguage {
    AR
    DE
    EN
    ES
    FR
    HE
    IT
    NL
    NO
    PT
    RU
    SE
    UD
    ZH
  }

  type Articles {
    items: [Article!]!
    hasMore: Boolean!
  }

  type ArticlesResult {
    items: [ArticleResponse!]!
    hasMore: Boolean!
  }

  type Source {
    name: String
    id: String
  }

  type ArticleResponse {
    description: String
    content: String
    urlToImage: String
    author: String
    publishedAt: String
    source: Source!
    url: String
    title: String
  }

  type Article {
    publishedAt: String
    description: String
    content: String
    source: String
    author: String
    title: String
    image: String
    url: String
    id: ID
  }
`;
