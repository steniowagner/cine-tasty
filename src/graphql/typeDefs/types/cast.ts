import { gql } from 'apollo-server';

export default gql`
  interface Cast {
    character: String
    backdrop_path: String
    overview: String
    vote_average: Float
    media_type: String
    poster_path: String
    popularity: Float
    original_language: String
    genre_ids(language: ISO6391Language): [String!]!
    vote_count: Float
    credit_id: String
    id: Int
  }

  type CastMovie implements Cast {
    original_title: String
    video: Boolean
    title: String
    adult: Boolean
    release_date: String
    character: String
    backdrop_path: String
    genre_ids(language: ISO6391Language): [String!]!
    overview: String
    vote_average: Float
    media_type: String
    poster_path: String
    popularity: Float
    original_language: String
    vote_count: Float
    credit_id: String
    id: Int
  }

  type CastTVShow implements Cast {
    episode_count: Int
    origin_country: [String!]!
    original_name: String
    name: String
    first_air_date: String
    character: String
    backdrop_path: String
    genre_ids(language: ISO6391Language): [String!]!
    overview: String
    vote_average: Float
    media_type: String
    poster_path: String
    popularity: Float
    original_language: String
    vote_count: Float
    credit_id: String
    id: Int
  }

  type CastItem {
    name: String
    profilePath: String
    id: ID
    character: String
    gender: Int
    order: Int
  }

  type CastItemResponse {
    name: String
    profile_path: String
    id: ID
    character: String
    gender: Int
    order: Int
  }

  type CrewItem {
    department: String
    id: ID
    job: String
    name: String
    gender: Int
    profilePath: String
  }

  type CrewItemResponse {
    department: String
    id: ID
    job: String
    name: String
    gender: Int
    profile_path: String
  }
`;
