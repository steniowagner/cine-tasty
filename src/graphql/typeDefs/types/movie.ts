import { gql } from 'apollo-server';

export default gql`
  type Movie {
    adult: Boolean
    backdropPath: String
    genres(language: ISO6391Language): [String!]!
    id: ID
    originalLanguage: String
    originalTitle: String
    overview: String
    posterPath: String
    popularity: Float
    video: Boolean
    title: String
    voteAverage: Float
    releaseDate: String
    productionCompanies: [ProductionCompany!]!
    voteCount: Float
    runtime: Float
    spokenLanguages: [String!]!
    status: String
    tagline: String
    budget: Float
    homepage: String
    revenue: Float
    productionCountries: [String!]!
    cast: [CastItem!]!
    crew: [CrewItem!]!
    videos: [MediaVideo!]!
    images(id: ID!): [String!]!
    similar: [BaseMovie!]!
    reviews: [Review!]!
  }

  type MovieResponse {
    adult: Boolean
    backdrop_path: String
    genres(language: ISO6391Language): [String!]!
    id: ID
    original_language: String
    original_title: String
    overview: String
    poster_path: String
    popularity: Float
    video: Boolean
    title: String
    vote_average: Float
    release_date: String
    production_companies: [ProductionCompanyResponse!]!
    vote_count: Float
    runtime: Float
    spoken_languages: [String!]!
    status: String
    tagline: String
    budget: Float
    homepage: String
    revenue: Float
    production_countries: [String!]!
    cast: [CastItemResponse!]!
    crew: [CrewItemResponse!]!
    videos: [MediaVideo!]!
    images(id: ID!): [String!]!
    similar: [BaseMovieResponse!]!
    reviews: [Review!]!
  }

  type BaseMovie {
    originalTitle: String
    video: Boolean
    title: String
    adult: Boolean
    releaseDate: String
    backdropPath: String
    genreIds(language: ISO6391Language): [String!]!
    overview: String
    voteAverage: Float
    mediaType: String
    posterPath: String
    popularity: Float
    originalLanguage: String
    voteCount: Int
    id: Int
  }

  type BaseMovieResponse {
    original_title: String
    video: Boolean
    title: String
    adult: Boolean
    release_date: String
    backdrop_path: String
    genre_ids(language: ISO6391Language): [String!]!
    overview: String
    vote_average: Float
    media_type: String
    poster_path: String
    popularity: Float
    original_language: String
    vote_count: Int
    id: Int
  }

  input TrendingMoviesArgs {
    language: ISO6391Language
    page: Int!
  }

  type TrendingMoviesQueryResult {
    total_results: Int!
    total_pages: Int!
    items: [BaseMovie!]!
    hasMore: Boolean!
  }

  type TrendingMovies {
    now_playing(args: TrendingMoviesArgs!): TrendingMoviesQueryResult!
    popular(args: TrendingMoviesArgs!): TrendingMoviesQueryResult!
    top_rated(args: TrendingMoviesArgs!): TrendingMoviesQueryResult!
    upcoming(args: TrendingMoviesArgs!): TrendingMoviesQueryResult!
  }
`;
