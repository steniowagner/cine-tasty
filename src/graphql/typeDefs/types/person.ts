import { gql } from 'apollo-server';

export default gql`
  union PersonKnowFor = BaseMovie | BaseTVShow

  type PersonResponse {
    birthday: String
    known_for_department: String
    deathday: String
    id: Int
    name: String
    also_known_as: [String!]!
    place_of_birth: String
    profile_path: String
    adult: Boolean
    imdb_id: String
    homepage: String
    biography: String
    popularity: Float
    images: [String!]!
    gender: Int
    cast: [CastResponse!]!
  }

  type Person {
    birthday: String
    knownForDepartment: String
    deathday: String
    id: Int
    name: String
    alsoKnownAs: [String!]!
    placeOfBirth: String
    profilePath: String
    adult: Boolean
    imdbId: String
    homepage: String
    biography: String
    popularity: Float
    images: [String!]!
    gender: Int
    moviesCast: [CastMovie!]!
    tvCast: [CastTVShow!]!
  }

  type BasePersonResponse {
    profile_path: String
    adult: Boolean
    id: Int
    popularity: Float
    known_for: [PersonKnowFor!]!
    name: String
  }

  type BasePerson {
    profilePath: String
    adult: Boolean
    id: Int
    popularity: Float
    knownFor: [PersonKnowFor!]!
    name: String
  }

  type PeopleQueryResultResponse {
    total_results: Int!
    total_pages: Int!
    items: [BasePerson!]!
    hasMore: Boolean!
  }

  type PeopleQueryResult {
    totalResults: Int!
    totalPages: Int!
    items: [BasePerson!]!
    hasMore: Boolean!
  }
`;
