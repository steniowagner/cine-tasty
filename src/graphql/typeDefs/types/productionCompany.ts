import { gql } from 'apollo-server';

export default gql`
  type ProductionCompany {
    id: ID
    logoPath: String
    name: String
    originCountry: String
  }

  type ProductionCompanyResponse {
    id: ID
    logo_path: String
    name: String
    origin_country: String
  }
`;
