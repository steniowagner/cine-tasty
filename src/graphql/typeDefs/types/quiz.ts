import { gql } from 'apollo-server';

export default gql`
  enum QuestionDifficulty {
    EASY
    MEDIUM
    HARD
    MIXED
  }

  enum QuestionType {
    MULTIPLE
    BOOLEAN
    MIXED
  }

  enum QuestionCategory {
    MOVIE
    TV
    MIXED
  }

  type Question {
    options: [String!]!
    category: String!
    type: String!
    difficulty: String!
    question: String!
    correctAnswer: String!
  }

  input QuizInput {
    difficulty: QuestionDifficulty!
    type: QuestionType!
    category: QuestionCategory!
    numberOfQuestions: Int!
  }
`;
