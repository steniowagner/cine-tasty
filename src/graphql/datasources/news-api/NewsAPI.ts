import { RESTDataSource } from 'apollo-datasource-rest';

import { ArticleQueryResult, QueryArticlesArgs } from '@lib/types';
import { GetArticlesResultItem } from '@news-api-types';

import { makeRequestParams } from './helpers';
import CONSTANTS from './utils/constants';

type GetRequestResponse = {
  articles: GetArticlesResultItem[];
  status: string;
};

class NewsAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = CONSTANTS.BASE_URL;
  }

  async getArticles({ page, language }: QueryArticlesArgs): Promise<ArticleQueryResult> {
    const params = makeRequestParams(page, language);

    try {
      const { articles } = await this.get<GetRequestResponse>(CONSTANTS.ENDPOINT, params);

      return {
        hasMore: articles.length === CONSTANTS.PAGE_SIZE,
        items: articles,
      };
    } catch (err) {
      return {
        hasMore: false,
        items: [],
      };
    }
  }
}

export default NewsAPI;
