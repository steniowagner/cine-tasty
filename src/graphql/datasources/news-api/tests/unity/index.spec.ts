const mockRestDataSourceGet = jest.fn();

import { ArticleLanguage } from '../../../../../lib/types';
import {
  rawArticleWithId,
  articleWithId,
} from '../../../../../__tests__/mocks/articles.stub';
import getDateParam from '../../helpers/getDateParam';
import env from '../../../../../config/environment';
import CONSTANTS from '../../utils/constants';
import NewsAPI from '../..';

const dateParam = getDateParam();

jest.mock('apollo-datasource-rest', () => {
  class MockRESTDataSource {
    baseUrl = '';
    get = mockRestDataSourceGet;
  }

  return {
    RESTDataSource: MockRESTDataSource,
    HTTPCache: class HTTPCache {},
  };
});

describe('Unity: NewsAPI', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('getArticles()', () => {
    it('should fetch movie articles from NewsAPI and return it correctly', async () => {
      mockRestDataSourceGet.mockReturnValue({
        status: CONSTANTS.STATUS_OK,
        articles: [rawArticleWithId],
      });

      const newsAPI = new NewsAPI();

      const result = await newsAPI.getArticles({ page: 1, language: ArticleLanguage.En });

      expect(mockRestDataSourceGet.mock.calls.length).toBe(1);

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(CONSTANTS.ENDPOINT, {
        language: ArticleLanguage.En.toLowerCase(),
        pageSize: CONSTANTS.PAGE_SIZE,
        apiKey: env.NEWS_API_KEY,
        q: CONSTANTS.QUERY,
        from: dateParam,
        to: dateParam,
        page: 1,
      });

      expect(result).toMatchSnapshot();
    });

    it('should fetch movie articles from NewsAPI and return it correctly when a language is provided', async () => {
      mockRestDataSourceGet.mockReturnValue({
        status: CONSTANTS.STATUS_OK,
        articles: [rawArticleWithId],
      });

      const newsAPI = new NewsAPI();

      const result = await newsAPI.getArticles({ page: 1, language: ArticleLanguage.Pt });

      expect(mockRestDataSourceGet.mock.calls.length).toBe(1);

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(CONSTANTS.ENDPOINT, {
        apiKey: env.NEWS_API_KEY,
        pageSize: CONSTANTS.PAGE_SIZE,
        from: dateParam,
        language: 'pt',
        to: dateParam,
        q: CONSTANTS.QUERY,
        page: 1,
      });

      expect(result).toEqual({
        hasMore: false,
        items: [articleWithId],
      });
    });

    it('should return an empty array when the returned status is other than ok', async () => {
      mockRestDataSourceGet.mockReturnValue({
        status: 'other',
      });

      const newsAPI = new NewsAPI();

      const result = await newsAPI.getArticles({ page: 1, language: ArticleLanguage.En });

      expect(mockRestDataSourceGet.mock.calls.length).toBe(1);

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(CONSTANTS.ENDPOINT, {
        language: ArticleLanguage.En.toLowerCase(),
        pageSize: CONSTANTS.PAGE_SIZE,
        apiKey: env.NEWS_API_KEY,
        q: CONSTANTS.QUERY,
        from: dateParam,
        to: dateParam,
        page: 1,
      });

      expect(result).toEqual({
        hasMore: false,
        items: [],
      });
    });
  });
});
