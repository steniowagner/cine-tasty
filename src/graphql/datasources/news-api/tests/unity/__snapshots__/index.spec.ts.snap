// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unity: NewsAPI getArticles() should fetch movie articles from NewsAPI and return it correctly 1`] = `
Object {
  "hasMore": false,
  "items": Array [
    Object {
      "author": "author",
      "content": "content",
      "description": undefined,
      "id": "url",
      "image": "urlToImage",
      "publishedAt": "publishedAt",
      "source": "name",
      "title": "title",
      "url": "url",
    },
  ],
}
`;
